File: internal/subscription/parser.go
package subscription

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/url"
	"strconv"
	"strings"

	"gopkg.in/yaml.v3"
)

func Parse(text string) ParseResult {
	trimmed := strings.TrimSpace(text)
	if trimmed == "" {
		return ParseResult{Format: "unknown", Nodes: []Proxy{}, DiscardedPlaceholders: []Proxy{}, Diagnostics: []string{"Response body is empty"}}
	}
	yamlHint := strings.HasPrefix(trimmed, "proxies:") || strings.HasPrefix(trimmed, "port:") || strings.HasPrefix(trimmed, "#") || strings.Contains(trimmed, "\nproxies:")
	result := ParseResult{Nodes: []Proxy{}, DiscardedPlaceholders: []Proxy{}, Diagnostics: []string{}}
	if decoded, ok := decodeSubscription(trimmed); ok && containsProxyURI(string(decoded)) && !yamlHint {
		result.Format = "base64"
		result.DecodedText = string(decoded)
		for index, line := range strings.Split(string(decoded), "\n") {
			line = strings.TrimSpace(line)
			if line == "" {
				continue
			}
			proxy, err := ParseURI(line)
			if err != nil {
				scheme, _, _ := strings.Cut(line, "://")
				if scheme == "" {
					scheme = "unknown"
				}
				result.Diagnostics = append(result.Diagnostics, fmt.Sprintf("Line %d uses an unsupported or invalid proxy URI (%s)", index+1, scheme))
				continue
			}
			result.Nodes = append(result.Nodes, proxy)
		}
	} else {
		result.Format = "yaml"
		var document struct {
			Proxies []map[string]any `yaml:"proxies"`
		}
		if err := yaml.Unmarshal([]byte(text), &document); err != nil {
			if !yamlHint {
				result.Format = "unknown"
			}
			result.Diagnostics = append(result.Diagnostics, "YAML parse failed: "+err.Error())
			return result
		}
		for index, value := range document.Proxies {
			proxy, err := ProxyFromMap(value)
			if err != nil {
				result.Diagnostics = append(result.Diagnostics, fmt.Sprintf("Proxy %d is invalid: %v", index+1, err))
				continue
			}
			result.Nodes = append(result.Nodes, proxy)
		}
		if len(result.Nodes) == 0 {
			result.Diagnostics = append(result.Diagnostics, "YAML response contains no proxy nodes")
		}
	}
	usable := result.Nodes[:0]
	for _, proxy := range result.Nodes {
		if (proxy.Server == "127.0.0.1" || proxy.Server == "::1" || proxy.Server == "localhost") && proxy.Port <= 1 {
			result.DiscardedPlaceholders = append(result.DiscardedPlaceholders, proxy)
		} else {
			usable = append(usable, proxy)
		}
	}
	result.Nodes = usable
	if count := len(result.DiscardedPlaceholders); count > 0 {
		result.Diagnostics = append(result.Diagnostics, fmt.Sprintf("Discarded %d placeholder node(s) using a loopback address and port 0 or 1", count))
	}
	return result
}

func ProxyFromMap(value map[string]any) (Proxy, error) {
	proxy := Proxy{Extra: map[string]any{}}
	for key, item := range value {
		switch key {
		case "name":
			proxy.Name, _ = item.(string)
		case "type":
			proxy.Type, _ = item.(string)
		case "server":
			proxy.Server, _ = item.(string)
		case "port":
			proxy.Port = integer(item)
		default:
			proxy.Extra[key] = item
		}
	}
	proxy.Type = strings.ToLower(proxy.Type)
	if proxy.Name == "" || proxy.Type == "" || proxy.Server == "" || proxy.Port < 0 || proxy.Port > 65535 {
		return Proxy{}, fmt.Errorf("name, type, server, and a valid port are required")
	}
	return proxy, nil
}

func ParseURI(value string) (Proxy, error) {
	switch {
	case strings.HasPrefix(value, "vless://"):
		return parseVLESSURI(value)
	case strings.HasPrefix(value, "trojan://"):
		return parseTrojanURI(value)
	case strings.HasPrefix(value, "hysteria2://"), strings.HasPrefix(value, "hy2://"):
		return parseHysteria2URI(value)
	case strings.HasPrefix(value, "anytls://"):
		return parseAnyTLSURI(value)
	case strings.HasPrefix(value, "vmess://"):
		return parseVMessURI(value)
	case strings.HasPrefix(value, "ss://"):
		return parseSSURI(value)
	default:
		return Proxy{}, fmt.Errorf("unsupported proxy URI")
	}
}

func parseProxyURL(value, proxyType string) (*url.URL, string, int, error) {
	parsed, err := url.Parse(value)
	if err != nil || parsed.Hostname() == "" || parsed.Port() == "" {
		return nil, "", 0, fmt.Errorf("invalid %s URI", proxyType)
	}
	port, err := strconv.Atoi(parsed.Port())
	if err != nil || port < 0 || port > 65535 {
		return nil, "", 0, fmt.Errorf("invalid port")
	}
	name, _ := url.PathUnescape(parsed.Fragment)
	if name == "" {
		name = fmt.Sprintf("%s:%d", parsed.Hostname(), port)
	}
	return parsed, name, port, nil
}

func parseVLESSURI(value string) (Proxy, error) {
	parsed, name, port, err := parseProxyURL(value, "vless")
	if err != nil {
		return Proxy{}, err
	}
	uuid := ""
	if parsed.User != nil {
		uuid = parsed.User.Username()
	}
	extra := map[string]any{"uuid": uuid, "udp": true}
	query := parsed.Query()
	network := valueOr(query.Get("type"), "tcp")
	if network != "tcp" {
		extra["network"] = network
	}
	if network == "ws" {
		options := map[string]any{"path": valueOr(query.Get("path"), "/")}
		if host := query.Get("host"); host != "" {
			options["headers"] = map[string]any{"Host": host}
		}
		extra["ws-opts"] = options
	}
	if network == "grpc" {
		extra["grpc-opts"] = map[string]any{"grpc-service-name": query.Get("serviceName")}
	}
	security := query.Get("security")
	if security == "tls" {
		extra["tls"] = true
		if sni := query.Get("sni"); sni != "" {
			extra["servername"] = sni
		}
		if fingerprint := query.Get("fp"); fingerprint != "" {
			extra["client-fingerprint"] = fingerprint
		}
		if alpn := query.Get("alpn"); alpn != "" {
			extra["alpn"] = strings.Split(alpn, ",")
		}
		if query.Get("insecure") == "1" {
			extra["skip-cert-verify"] = true
		}
	}
	if security == "reality" {
		extra["tls"] = true
		extra["reality-opts"] = map[string]any{"public-key": query.Get("pbk"), "short-id": query.Get("sid")}
		if sni := query.Get("sni"); sni != "" {
			extra["servername"] = sni
		}
		if fingerprint := query.Get("fp"); fingerprint != "" {
			extra["client-fingerprint"] = fingerprint
		}
	}
	if flow := query.Get("flow"); flow != "" {
		extra["flow"] = flow
	}
	return Proxy{Name: name, Type: "vless", Server: parsed.Hostname(), Port: port, Extra: extra}, nil
}

func parseTrojanURI(value string) (Proxy, error) {
	parsed, name, port, err := parseProxyURL(value, "trojan")
	if err != nil {
		return Proxy{}, err
	}
	password := ""
	if parsed.User != nil {
		password = parsed.User.Username()
	}
	extra := map[string]any{"password": password, "udp": true}
	query := parsed.Query()
	if sni := query.Get("sni"); sni != "" {
		extra["sni"] = sni
	}
	if alpn := query.Get("alpn"); alpn != "" {
		extra["alpn"] = strings.Split(alpn, ",")
	}
	if fingerprint := query.Get("fp"); fingerprint != "" {
		extra["client-fingerprint"] = fingerprint
	}
	if query.Get("insecure") == "1" || query.Get("allowInsecure") == "1" {
		extra["skip-cert-verify"] = true
	}
	network := valueOr(query.Get("type"), "tcp")
	if network == "ws" || network == "grpc" {
		extra["network"] = network
	}
	if network == "ws" {
		options := map[string]any{"path": valueOr(query.Get("path"), "/")}
		if host := query.Get("host"); host != "" {
			options["headers"] = map[string]any{"Host": host}
		}
		extra["ws-opts"] = options
	}
	if network == "grpc" {
		extra["grpc-opts"] = map[string]any{"grpc-service-name": query.Get("serviceName")}
	}
	return Proxy{Name: name, Type: "trojan", Server: parsed.Hostname(), Port: port, Extra: extra}, nil
}

func parseHysteria2URI(value string) (Proxy, error) {
	parsed, name, port, err := parseProxyURL(value, "hysteria2")
	if err != nil {
		return Proxy{}, err
	}
	password := ""
	if parsed.User != nil {
		password = parsed.User.Username()
	}
	extra := map[string]any{"password": password}
	query := parsed.Query()
	if sni := query.Get("sni"); sni != "" {
		extra["sni"] = sni
	}
	if obfs, obfsPassword := query.Get("obfs"), query.Get("obfs-password"); obfs != "" && obfsPassword != "" {
		extra["obfs"] = obfs
		extra["obfs-password"] = obfsPassword
	}
	if query.Get("insecure") == "1" {
		extra["skip-cert-verify"] = true
	}
	if alpn := query.Get("alpn"); alpn != "" {
		extra["alpn"] = strings.Split(alpn, ",")
	}
	return Proxy{Name: name, Type: "hysteria2", Server: parsed.Hostname(), Port: port, Extra: extra}, nil
}

func parseAnyTLSURI(value string) (Proxy, error) {
	parsed, name, port, err := parseProxyURL(value, "anytls")
	if err != nil {
		return Proxy{}, err
	}
	password := ""
	if parsed.User != nil {
		password = parsed.User.Username()
	}
	extra := map[string]any{"password": password, "udp": true}
	query := parsed.Query()
	if sni := query.Get("sni"); sni != "" {
		extra["sni"] = sni
	}
	if query.Get("insecure") == "1" {
		extra["skip-cert-verify"] = true
	}
	if fingerprint := query.Get("fp"); fingerprint != "" {
		extra["client-fingerprint"] = fingerprint
	}
	if alpn := query.Get("alpn"); alpn != "" {
		extra["alpn"] = strings.Split(alpn, ",")
	}
	return Proxy{Name: name, Type: "anytls", Server: parsed.Hostname(), Port: port, Extra: extra}, nil
}

func parseVMessURI(value string) (Proxy, error) {
	content := strings.TrimPrefix(value, "vmess://")
	content, _, _ = strings.Cut(content, "#")
	if decoded, ok := decodeSubscription(content); ok {
		var object map[string]any
		if json.Unmarshal(decoded, &object) == nil {
			server, _ := object["add"].(string)
			port := integer(object["port"])
			if server != "" && port > 0 && port <= 65535 {
				name, _ := object["ps"].(string)
				if name == "" {
					name, _ = object["remarks"].(string)
				}
				if name == "" {
					name = fmt.Sprintf("%s:%d", server, port)
				}
				extra := map[string]any{"alterId": integer(object["aid"]), "cipher": valueOr(stringValue(object["scy"]), "auto"), "udp": true}
				if id := stringValue(object["id"]); id != "" {
					extra["uuid"] = id
				}
				network := valueOr(stringValue(object["net"]), "tcp")
				if network != "tcp" {
					extra["network"] = network
				}
				if network == "ws" {
					opts := map[string]any{"path": valueOr(stringValue(object["path"]), "/")}
					if host := stringValue(object["host"]); host != "" {
						opts["headers"] = map[string]any{"Host": host}
					}
					extra["ws-opts"] = opts
				}
				if network == "grpc" {
					extra["grpc-opts"] = map[string]any{"grpc-service-name": stringValue(object["path"])}
				}
				if stringValue(object["tls"]) == "tls" {
					extra["tls"] = true
					copyObject(extra, object, "sni", "servername", "fp", "client-fingerprint")
					switch alpn := object["alpn"].(type) {
					case string:
						if alpn != "" {
							extra["alpn"] = strings.Split(alpn, ",")
						}
					case []any:
						extra["alpn"] = alpn
					}
				}
				return Proxy{Name: name, Type: "vmess", Server: server, Port: port, Extra: extra}, nil
			}
		}
	}
	parsed, name, port, err := parseProxyURL(value, "vmess")
	if err != nil {
		return Proxy{}, err
	}
	uuid := ""
	if parsed.User != nil {
		uuid = parsed.User.Username()
	}
	return Proxy{Name: name, Type: "vmess", Server: parsed.Hostname(), Port: port, Extra: map[string]any{"uuid": uuid, "alterId": 0, "cipher": "auto", "udp": true}}, nil
}

func parseSSURI(value string) (Proxy, error) {
	parsed, err := url.Parse(value)
	if err == nil && parsed.Hostname() != "" && parsed.Port() != "" && parsed.User != nil {
		port, portErr := strconv.Atoi(parsed.Port())
		if portErr != nil || port < 0 || port > 65535 {
			return Proxy{}, fmt.Errorf("invalid ss URI")
		}
		userinfo := parsed.User.Username()
		decoded, ok := decodeSubscription(userinfo)
		if ok {
			method, password, found := strings.Cut(string(decoded), ":")
			if found {
				name, _ := url.PathUnescape(parsed.Fragment)
				if name == "" {
					name = fmt.Sprintf("%s:%d", parsed.Hostname(), port)
				}
				extra := map[string]any{"cipher": method, "password": password, "udp": true}
				if plugin := parsed.Query().Get("plugin"); plugin != "" {
					parseSSPlugin(plugin, extra)
				}
				return Proxy{Name: name, Type: "ss", Server: parsed.Hostname(), Port: port, Extra: extra}, nil
			}
		}
	}
	raw := strings.TrimPrefix(value, "ss://")
	raw, fragment, _ := strings.Cut(raw, "#")
	raw, _, _ = strings.Cut(raw, "?")
	decoded, ok := decodeSubscription(raw)
	if !ok {
		return Proxy{}, fmt.Errorf("invalid ss URI")
	}
	decodedText := string(decoded)
	at := strings.LastIndex(decodedText, "@")
	if at < 0 {
		return Proxy{}, fmt.Errorf("invalid ss URI")
	}
	userinfo, address := decodedText[:at], decodedText[at+1:]
	method, password, credentials := strings.Cut(userinfo, ":")
	colon := strings.LastIndex(address, ":")
	if colon < 0 {
		return Proxy{}, fmt.Errorf("invalid ss URI")
	}
	host, portText := address[:colon], address[colon+1:]
	port, portErr := strconv.Atoi(portText)
	if !credentials || host == "" || portErr != nil || port < 0 || port > 65535 {
		return Proxy{}, fmt.Errorf("invalid ss URI")
	}
	name, _ := url.PathUnescape(fragment)
	if name == "" {
		name = fmt.Sprintf("%s:%d", host, port)
	}
	return Proxy{Name: name, Type: "ss", Server: host, Port: port, Extra: map[string]any{"cipher": method, "password": password, "udp": true}}, nil
}

func parseSSPlugin(plugin string, extra map[string]any) {
	parts := strings.Split(plugin, ";")
	extra["plugin"] = parts[0]
	options := map[string]any{}
	for _, part := range parts[1:] {
		key, value, ok := strings.Cut(part, "=")
		if ok {
			options[key] = value
		}
	}
	if parts[0] == "obfs-local" || parts[0] == "obfs" {
		if value, ok := options["obfs"]; ok {
			options["mode"] = value
			delete(options, "obfs")
		}
		if value, ok := options["obfs-host"]; ok {
			options["host"] = value
			delete(options, "obfs-host")
		}
	}
	if len(options) > 0 {
		extra["plugin-opts"] = options
	}
}

func decodeSubscription(value string) ([]byte, bool) {
	for _, encoding := range []*base64.Encoding{base64.StdEncoding, base64.RawStdEncoding, base64.URLEncoding, base64.RawURLEncoding} {
		if decoded, err := encoding.DecodeString(strings.TrimSpace(value)); err == nil {
			return decoded, true
		}
	}
	return nil, false
}

func containsProxyURI(value string) bool {
	for _, line := range strings.Split(value, "\n") {
		line = strings.TrimSpace(line)
		for _, prefix := range []string{"vless://", "vmess://", "ss://", "trojan://", "ssr://", "hysteria://", "hysteria2://", "hy2://", "anytls://"} {
			if strings.HasPrefix(line, prefix) {
				return true
			}
		}
	}
	return false
}

func integer(value any) int {
	switch item := value.(type) {
	case int:
		return item
	case int64:
		return int(item)
	case uint64:
		return int(item)
	case float64:
		return int(item)
	case string:
		result, _ := strconv.Atoi(item)
		return result
	}
	return 0
}
func stringValue(value any) string { result, _ := value.(string); return result }
func valueOr(value, fallback string) string {
	if value == "" {
		return fallback
	}
	return value
}
func copyObject(target, source map[string]any, pairs ...string) {
	for index := 0; index+1 < len(pairs); index += 2 {
		if value := source[pairs[index]]; value != nil && value != "" {
			target[pairs[index+1]] = value
		}
	}
}

File: rust/crates/sempre-converter/src/parser.rs
mod uri;

use base64::Engine as _;
use base64::engine::general_purpose::{STANDARD, URL_SAFE_NO_PAD};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::Proxy;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParseResult {
    pub format: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub decoded_text: String,
    pub nodes: Vec<Proxy>,
    pub discarded_placeholder_nodes: Vec<Proxy>,
    pub diagnostics: Vec<String>,
}

impl ParseResult {
    pub(crate) fn messages(&self) -> &[String] {
        &self.diagnostics
    }
}

pub fn parse_subscription(text: &str) -> ParseResult {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        return result("unknown", vec![], vec!["Response body is empty".into()]);
    }
    let yaml_hint = trimmed.starts_with("proxies:")
        || trimmed.starts_with("port:")
        || trimmed.starts_with('#')
        || trimmed.contains("\nproxies:");
    let decoded =
        decode_subscription(trimmed).filter(|value| contains_proxy_uri(value) && !yaml_hint);
    let mut parsed = if let Some(decoded) = decoded {
        parse_uri_lines("base64", decoded)
    } else if contains_proxy_uri(trimmed) && !yaml_hint {
        parse_uri_lines("uri-list", trimmed.into())
    } else {
        parse_yaml(text, yaml_hint)
    };
    discard_placeholders(&mut parsed);
    parsed
}

fn parse_uri_lines(format: &str, text: String) -> ParseResult {
    let mut nodes = Vec::new();
    let mut diagnostics = Vec::new();
    for (index, line) in text.lines().map(str::trim).enumerate() {
        if line.is_empty() {
            continue;
        }
        match uri::parse(line) {
            Ok(proxy) => nodes.push(proxy),
            Err(error) => diagnostics.push(format!(
                "Line {} uses an unsupported or invalid proxy URI: {error}",
                index + 1
            )),
        }
    }
    let decoded_text = if format == "base64" {
        text
    } else {
        String::new()
    };
    ParseResult {
        format: format.into(),
        decoded_text,
        nodes,
        discarded_placeholder_nodes: Vec::new(),
        diagnostics,
    }
}

fn parse_yaml(text: &str, yaml_hint: bool) -> ParseResult {
    let decoded: Value = match serde_yaml::from_str(text) {
        Ok(value) => value,
        Err(error) => {
            return result(
                if yaml_hint { "yaml" } else { "unknown" },
                vec![],
                vec![format!("YAML parse failed: {error}")],
            );
        }
    };
    let values = decoded
        .get("proxies")
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    let mut nodes = Vec::new();
    let mut diagnostics = Vec::new();
    for (index, value) in values.into_iter().enumerate() {
        match Proxy::from_value(value) {
            Ok(proxy) if valid_proxy(&proxy) => nodes.push(proxy),
            Ok(_) | Err(_) => diagnostics.push(format!(
                "Proxy {} is invalid: name, type, server, and a valid port are required",
                index + 1
            )),
        }
    }
    if nodes.is_empty() {
        diagnostics.push("YAML response contains no proxy nodes".into());
    }
    result("yaml", nodes, diagnostics)
}

fn valid_proxy(proxy: &Proxy) -> bool {
    !proxy.name.trim().is_empty()
        && !proxy.proxy_type.trim().is_empty()
        && !proxy.server.trim().is_empty()
        && proxy.port > 0
}

fn discard_placeholders(parsed: &mut ParseResult) {
    let mut usable = Vec::with_capacity(parsed.nodes.len());
    for proxy in parsed.nodes.drain(..) {
        if matches!(proxy.server.as_str(), "127.0.0.1" | "::1" | "localhost") && proxy.port <= 1 {
            parsed.discarded_placeholder_nodes.push(proxy);
        } else {
            usable.push(proxy);
        }
    }
    parsed.nodes = usable;
    if !parsed.discarded_placeholder_nodes.is_empty() {
        parsed.diagnostics.push(format!(
            "Discarded {} placeholder node(s) using a loopback address and port 0 or 1",
            parsed.discarded_placeholder_nodes.len()
        ));
    }
}

fn result(format: &str, nodes: Vec<Proxy>, diagnostics: Vec<String>) -> ParseResult {
    ParseResult {
        format: format.into(),
        decoded_text: String::new(),
        nodes,
        discarded_placeholder_nodes: Vec::new(),
        diagnostics,
    }
}

fn contains_proxy_uri(value: &str) -> bool {
    [
        "vless://",
        "vmess://",
        "ss://",
        "trojan://",
        "hysteria2://",
        "hy2://",
        "anytls://",
    ]
    .iter()
    .any(|scheme| value.contains(scheme))
}

fn decode_subscription(value: &str) -> Option<String> {
    let compact: String = value
        .chars()
        .filter(|character| !character.is_whitespace())
        .collect();
    for engine in [&STANDARD, &URL_SAFE_NO_PAD] {
        let mut padded = compact.clone();
        while !padded.len().is_multiple_of(4) {
            padded.push('=');
        }
        if let Ok(bytes) = engine.decode(padded.as_bytes())
            && let Ok(text) = String::from_utf8(bytes)
        {
            return Some(text);
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::parse_subscription;

    #[test]
    fn parses_yaml_and_discards_placeholder() {
        let parsed = parse_subscription(
            "proxies:\n  - { name: edge, type: socks5, server: edge.example.com, port: 1080 }\n  - { name: empty, type: socks5, server: 127.0.0.1, port: 1 }",
        );
        assert_eq!(parsed.nodes.len(), 1);
        assert_eq!(parsed.discarded_placeholder_nodes.len(), 1);
    }

    #[test]
    fn parses_plain_uri_list() {
        let parsed = parse_subscription(
            "vless://id@example.com:443?security=tls&sni=edge.example.com#Hong%20Kong%20%E9%A6%99%E6%B8%AF",
        );
        assert_eq!(parsed.nodes[0].name, "Hong Kong 香港");
        assert_eq!(parsed.nodes[0].proxy_type, "vless");
    }
}
